File: 03/claude-opus-4.5/python/solution.py
with open("input.txt") as f:
    lines = [line.strip() for line in f if line.strip()]

def max_joltage(line: str) -> int:
    n = len(line)
    if n < 2:
        return 0

    # Precompute suffix maximum
    suffix_max = [0] * n
    for i in range(n - 2, -1, -1):
        digit = int(line[i + 1])
        suffix_max[i] = max(digit, suffix_max[i + 1])

    # Find max joltage
    max_j = 0
    for i in range(n - 1):
        tens_digit = int(line[i])
        units_digit = suffix_max[i]
        joltage = tens_digit * 10 + units_digit
        max_j = max(max_j, joltage)

    return max_j

def max_joltage_part2(line: str, k: int = 12) -> int:
    n = len(line)
    if n < k:
        return 0
    if n == k:
        return int(line)

    # Greedy: pick k digits to form largest number
    result = []
    start_idx = 0

    for i in range(k):
        remaining = k - i
        end_idx = n - remaining

        max_digit = '0'
        max_pos = start_idx

        for j in range(start_idx, end_idx + 1):
            if line[j] > max_digit:
                max_digit = line[j]
                max_pos = j

        result.append(max_digit)
        start_idx = max_pos + 1

    return int(''.join(result))

# Part 1
part1 = sum(max_joltage(line) for line in lines)
print(f"Day 03 Part 1: {part1}")

# Part 2
part2 = sum(max_joltage_part2(line) for line in lines)
print(f"Day 03 Part 2: {part2}")

File: 03/claude-opus-4.5/typescript/solution.ts
import * as fs from 'fs';

const input = fs.readFileSync('input.txt', 'utf-8').trim();
const lines = input.split('\n').filter(l => l.trim() !== '');

function maxJoltage(line: string): number {
    const n = line.length;
    if (n < 2) return 0;

    // Precompute suffix maximum
    const suffixMax: number[] = new Array(n).fill(0);
    for (let i = n - 2; i >= 0; i--) {
        const digit = parseInt(line[i + 1]);
        suffixMax[i] = Math.max(digit, suffixMax[i + 1]);
    }

    // Find max joltage
    let maxJ = 0;
    for (let i = 0; i < n - 1; i++) {
        const tensDigit = parseInt(line[i]);
        const unitsDigit = suffixMax[i];
        const joltage = tensDigit * 10 + unitsDigit;
        maxJ = Math.max(maxJ, joltage);
    }

    return maxJ;
}

function maxJoltagePart2(line: string, k: number = 12): bigint {
    const n = line.length;
    if (n < k) return 0n;
    if (n === k) return BigInt(line);

    // Greedy: pick k digits to form largest number
    const result: string[] = [];
    let startIdx = 0;

    for (let i = 0; i < k; i++) {
        const remaining = k - i;
        const endIdx = n - remaining;

        let maxDigit = '0';
        let maxPos = startIdx;

        for (let j = startIdx; j <= endIdx; j++) {
            if (line[j] > maxDigit) {
                maxDigit = line[j];
                maxPos = j;
            }
        }

        result.push(maxDigit);
        startIdx = maxPos + 1;
    }

    return BigInt(result.join(''));
}

// Part 1
let part1 = 0;
for (const line of lines) {
    part1 += maxJoltage(line.trim());
}
console.log(`Day 03 Part 1: ${part1}`);

// Part 2
let part2 = 0n;
for (const line of lines) {
    part2 += maxJoltagePart2(line.trim());
}
console.log(`Day 03 Part 2: ${part2}`);
